tests/async_action.cpp

```diff
@@ -99,4 +100,79 @@ TEST(async_action, time_invalid_format)
   EXPECT_THAT(log, testing::HasSubstr("strftime returned 0"));
 }
 
+TEST(async_action, helpererror)
+{
+  struct TestCase {
+    int func_id;
+    int return_value;
+    std::string expected_substring;
+    std::string filename;
+    unsigned int line;
+    unsigned int column;
+  };
+
+  std::vector<TestCase> test_cases = {
+    // case 1: `map_update_elem` returns `-E2BIG`
+    { .func_id = libbpf::BPF_FUNC_map_update_elem,
+      .return_value = -E2BIG,
+      .expected_substring = "WARNING: Map full; can't update element",
+      .filename = std::string("test1.bt"),
+      .line = 10,
+      .column = 5 },
+    // case 2: `map_delete_elem` returns `-ENOENT`
+    { .func_id = libbpf::BPF_FUNC_map_delete_elem,
+      .return_value = -ENOENT,
+      .expected_substring =
+          "WARNING: Can't delete map element because it does not exist",
+      .filename = std::string("test2.bt"),
+      .line = 15,
+      .column = 8 },
+    // case 3: `map_lookup_elem` failed to lookup map element
+    { .func_id = libbpf::BPF_FUNC_map_lookup_elem,
+      .return_value = 0,
+      .expected_substring =
+          "WARNING: Can't lookup map element because it does not exist",
+      .filename = std::string("test3.bt"),
+      .line = 20,
+      .column = 3 },
+    // case 4: default case - other function ID and error code
+    { .func_id = libbpf::BPF_FUNC_trace_printk,
+      .return_value = -EPERM,
+      .expected_substring = "WARNING: " + std::string(strerror(EPERM)),
+      .filename = std::string("test4.bt"),
+      .line = 25,
+      .column = 1 }
+  };
+
+  for (const auto &tc : test_cases) {
+    std::stringstream out;
+    auto bpftrace = get_mock_bpftrace(out);
+
+    auto src_loc = ast::SourceLocation(
+        location(&tc.filename, tc.line, tc.column));
+    auto location_chain = std::make_shared<ast::LocationChain>(src_loc);
+    HelperErrorInfo info(tc.func_id, location_chain);
+
+    bpftrace->resources.helper_error_info.emplace(tc.func_id, std::move(info));
```

Comment:
This seems unnecessarily complicated with all the `make_shared` and `std::move`.

Wouldn't this be possible?
```suggestion
    auto src_loc = ast::SourceLocation(
        location(&tc.filename, tc.line, tc.column));
    auto location_chain = LocationChain(src_loc);

    bpftrace->resources.helper_error_info.emplace(tc.func_id, HelperErrorInfo info(tc.func_id, location_chain));
```

Reply:
That's because the `HelperErrorInfo` requires a `ast::Location` type parameter and the `ast::Location` is a type alias of `std::shared_ptr<LocationChain>`

FYI: 
https://github.com/bpftrace/bpftrace/blob/master/src/required_resources.h#L30-L41
https://github.com/bpftrace/bpftrace/blob/master/src/ast/location.h#L67